### PR TITLE
Allow calculated LLR values to have an interval (optional)

### PR DESCRIPTION
**Pull request by JoBraNFI**
## Scope

Calculated LLR values are now represented by an `LLRData` tuple, which may or may not contain intervals (indicated by two columns; low and high percentile ranges respectively).

### CODE REVIEW

File: lir/experiment.py [hunk 1]
```diff
@@ -9,12 +9,27 @@
...
+from lir.lrsystems.lrsystems import LRSystem, LLRData
 
 LOG = logging.getLogger(__name__)
 
 
+def combine_results(llr_sets: list[LLRData]) -> LLRData:
+    """Combine the results of the LLRData tuples into a single tuple."""
+
+    llrs = [llr_data_tuple.llrs for llr_data_tuple in llr_sets]
+    intervals = [llr_data_tuple.intervals for llr_data_tuple in llr_sets if llr_data_tuple.intervals is not None]
```
Comment (wowtor) [inline lir/experiment.py]: `if` clause weg hier bij intervals en labels (anders gaat het mis bij regel 27 en 28)

File: lir/lrsystems/lrsystems.py [hunk 2]
```diff
@@ -65,6 +65,13 @@ def fit_transform(
         return super().transform(features)
 
 
+class LLRData(NamedTuple):
```
Comment (wowtor) [inline lir/lrsystems/lrsystems.py:68]: misschien moeten we hier nog een docstr aan geven...

File: lir/lrsystems/lrsystems.py [hunk 3]
```diff
@@ -81,9 +88,7 @@ def fit(self, instances: np.ndarray, labels: np.ndarray, meta: np.ndarray) -> "L
         return self
 
     @abstractmethod
-    def apply(
-        self, instances: np.ndarray, labels: np.ndarray | None, meta: np.ndarray
-    ) -> tuple[np.ndarray, np.ndarray | None, np.ndarray]:
+    def apply(self, instances: np.ndarray, labels: np.ndarray | None, meta: np.ndarray) -> LLRData:
```
Comment (wowtor) [inline lir/lrsystems/lrsystems.py:97]: misschien kunnen we hier gelijk meenemen: instances -> features

Comment (wowtor) [inline lir/lrsystems/lrsystems.py:97]: docstr die de return type beschrijft past nu eigenlijk beter bij `LLRData`

**Review by wowtor — CHANGES_REQUESTED** [review]

File: lir/experiment.py (hunk 1 above)
Comment (JoBraNFI) [reply lir/experiment.py]: Oja, inderdaad. Thanks!

File: lir/lrsystems/lrsystems.py (hunk 3 above)
Comment (JoBraNFI) [reply lir/lrsystems/lrsystems.py:97]: dan moeten we dat op nog veel meer plekken hernoemen, lijkt me beter in een afzonderlijke PR

Comment (JoBraNFI) [reply lir/lrsystems/lrsystems.py:97]: Ik heb de docstrings nagelopen en aangepast (in de nieuwe commits)

File: lir/lrsystems/lrsystems.py (hunk 2 above)
Comment (JoBraNFI) [reply lir/lrsystems/lrsystems.py:68]: Goeie! Ik heb alle doc strings nog even nagelopen in de nieuwe commit

File: lir/lrsystems/lrsystems.py (hunk 3 above)
Comment (wowtor) [reply lir/lrsystems/lrsystems.py:97]: in andere files heet deze al `features`

File: lir/experiment.py [hunk 4]
```diff
@@ -2,19 +2,56 @@
...
+
+    This helper function either directly returns `None` if a collection contains a `None` value, or
+    it uses a workaround by using a list comprehension which filters out the `None` values (which are not
+    present anymore) to satisfy the type checker.
+    """
+    if None in optional_values:
+        return None
+
+    # Workaround for type checking optional None values
+    return np.concatenate([value for value in optional_values if value is not None])
```
Comment (PimMeulensteen) [inline lir/experiment.py]: Kan dit ook niet een generator zijn? 

`np.concatenate(value for value in optional_values if value is not None)`

File: lir/experiment.py [hunk 5]
```diff
@@ -2,19 +2,56 @@
...
+
+def combine_results(llr_sets: list[LLRData]) -> LLRData:
+    """Combine the results of the LLRData tuples into a single tuple."""
+
+    llrs = [llr_data_tuple.llrs for llr_data_tuple in llr_sets]
+    intervals = [llr_data_tuple.intervals for llr_data_tuple in llr_sets]
+    labels = [llr_data_tuple.labels for llr_data_tuple in llr_sets]
+    meta_data = [llr_data_tuple.meta_data for llr_data_tuple in llr_sets]
+
+    # The `np_concatenate_optional()` helper function ensures values do not contain `None` values
```
Comment (PimMeulensteen) [inline lir/experiment.py]: Dit voelt hier niet zo op zijn plek, dit lijkt mij meer passen in `np_concatenate_optional()`.

**Review by PimMeulensteen — APPROVED** [review]

**Comment by wowtor** [comment]
Ik heb nog een branch hiervan afgetakt die deze lijn nog wat verder doorzet in de vorm van een dataclass, zie andere PR.

File: lir/experiment.py (hunk 4 above)
Comment (JoBraNFI) [reply lir/experiment.py]: Dat werkt niet, `np.concatenate()` werkt enkel op _sequence_'s

File: lir/experiment.py (hunk 5 above)
Comment (JoBraNFI) [reply lir/experiment.py]: Daar heb je wel gelijk in, ik zal het weghalen en de PR bijwerken

**Comment by JoBraNFI** [comment]
At second thought, it feels that working around the typing issue with `np_concatenate_optional()` does not seem the right approach. Especially, since we're constructing a new list and performing work **just** to satisfy the type checker.

I've updated the PR to reflect a better approach, using a type cast.
